server/sonar-server/src/main/java/org/sonar/server/qualitygate/RegisterQualityGates.java

```diff
@@ -37,67 +48,205 @@
+  public RegisterQualityGates(DbClient dbClient, QualityGateConditionsUpdater qualityGateConditionsUpdater, QualityGates qualityGates) {
     this.dbClient = dbClient;
     this.qualityGateConditionsUpdater = qualityGateConditionsUpdater;
-    this.loadedTemplateDao = loadedTemplateDao;
     this.qualityGates = qualityGates;
+    this.qualityGateDao = dbClient.qualityGateDao();
+    this.qualityGateConditionDao = dbClient.gateConditionDao();
   }
 
   @Override
   public void start() {
     try (DbSession dbSession = dbClient.openSession(false)) {
-      if (shouldRegisterBuiltinQualityGate(dbSession)) {
-        createBuiltinQualityGate(dbSession);
-        registerBuiltinQualityGate(dbSession);
-        dbSession.commit();
+      QualityGateDto builtin = qualityGateDao.selectByName(dbSession, BUILTIN_QUALITY_GATE_NAME);
+      if (builtin == null) {
+        LOGGER.info("Built-in quality gate [{}] has been created", BUILTIN_QUALITY_GATE_NAME);
+        builtin = createQualityGate(dbSession, BUILTIN_QUALITY_GATE_NAME);
+        if (qualityGates.getDefault() == null) {
+          qualityGates.setDefault(dbSession, builtin.getId());
+        }
       }
+
+      updateQualityConditionsIfRequired(dbSession, builtin);
+
+      qualityGateDao.ensureOneBuiltInQualityGate(dbSession, BUILTIN_QUALITY_GATE_NAME);
+
+      dbSession.commit();
     }
   }
 
-  @Override
-  public void stop() {
-    // do nothing
-  }
+  private void updateQualityConditionsIfRequired(DbSession dbSession, QualityGateDto builtin) {
+    List<QualityGateCondition> qualityGateConditions = qualityGateConditionDao.selectForQualityGate(dbSession, builtin.getId())
+      .stream()
+      .map(dto -> {
+        MetricDto metricDto = dbClient.metricDao().selectById(dbSession, dto.getMetricId());
+        if (metricDto != null) {
+          dto.setMetricKey(metricDto.getKey());
+        }
+        return QualityGateCondition.from(dto);
+      })
+      .collect(MoreCollectors.toList());
+
+    // Delete old conditions
+    List<QualityGateCondition> qgConditionsToBeDeleted = new ArrayList<>(qualityGateConditions);
+    qgConditionsToBeDeleted.removeAll(QUALITY_GATE_CONDITIONS);
+    qgConditionsToBeDeleted.stream()
+      .forEach(qgc -> qualityGateConditionDao.delete(qgc.toQualityGateDto(builtin.getId()), dbSession));
 
-  private boolean shouldRegisterBuiltinQualityGate(DbSession dbSession) {
-    return loadedTemplateDao.countByTypeAndKey(LoadedTemplateDto.QUALITY_GATE_TYPE, BUILTIN_QUALITY_GATE, dbSession) == 0;
+    // Create new conditions
+    List<QualityGateCondition> qgConditionsToBeCreated = new ArrayList<>(QUALITY_GATE_CONDITIONS);
+    qgConditionsToBeCreated.removeAll(qualityGateConditions);
+    qgConditionsToBeCreated.stream()
+      .forEach(qgc ->
+        qualityGateConditionsUpdater.createCondition(dbSession, builtin.getId(), qgc.getMetricKey(), qgc.getOperator(), qgc.getWarningThreshold(),
+          qgc.getErrorThreshold(), qgc.getPeriod())
+      );
+
```

Comment:
I found the algo not that easy to read. I think I would have prefer something like : 
- Read all existing condition dtos
- If one condition on metric/leak doesn't exist in the expected list, delete it
- If one condition on metric/leak has been updated, fix it
- At the end, add missing conditions

Reply:
It's pretty hard to know if a condition has been updated ? 
How do you know since the metric may have been updated, or the operator, or the value (or worst, the 3 of them). I think it's almost impossible to *update* a condition.

From my point of view, this algorithm is simple : 
* Delete the conditions that does not exist in the "Sonar way" quality gate,
* Add the missing conditions from the "Sonar way" quality gate

Reply:
ok, I understand